tests/test_TFNetworkLayer.py

```diff
@@ -3065,6 +3065,266 @@ def mask_func(source, **kwargs):
+    src_sparse = _create_data(n_batch=n_batch, n_time=n_time, n_feature=n_feature, network=network)
+    src = _create_data(n_batch=n_batch, n_time=n_time, n_feature=1, network=network, sparse=False)
+
+    vocab = _construct_vocab(n_time * n_batch)
+    if with_prob:
+      import pickle as pkl
+      vocab_prob = {w: 1 / (1000 * n_time * n_batch) for w in vocab.keys()}
+      vocab_prob['8'] = 1
+      vocab_prob['2'] = 1
+      with open("/tmp/vocab_probs.pkl", 'wb') as f:
+        pkl.dump(vocab_prob, f)
+
+    # construct mask layer:
+    settings = {"name": "out",
+                "sources": [src_sparse],
+                "network": network,
+                "vocab": "/tmp/vocab.pkl",
+                "vocab_probs": "/tmp/vocab_probs.pkl" if with_prob else None,
+                "mask_prob": 0.2,
+                "mask_m_prob": 0.7,
+                "mask_r_prob": 0.1,
+                "random_seed": None}
+    mask_out_output = ProbabilisticMaskLayer.get_out_data_from_opts(**settings)
+    mask_out = ProbabilisticMaskLayer(output=mask_out_output, **settings)
+    labels = mask_out.get_sub_layer("labels")
+    positions = mask_out.get_sub_layer("positions")
+
+    # construct gather layer:
+    kwargs = {"name": "out",
+              "sources": [src, positions],
+              "network": network}
+    out_output = GatherPositionsLayer.get_out_data_from_opts(**kwargs)
+    out = GatherPositionsLayer(output=out_output, **kwargs)
+
+    to_inspect = [src.output.placeholder,
+                  src.output.get_sequence_lengths(),
+                  out.output.placeholder,
+                  out.output.get_sequence_lengths(),
+                  labels.output.placeholder,
+                  positions.output.placeholder,
+                  positions.output.get_sequence_lengths()]
+    results = session.run(to_inspect)
+    src_res, src_lens_res, out_res, out_lens_res, labels_res, positions_res, num_positions_res = results[:7]
+
+    # check the results:
+    print("Check if we recovered the labels correctly:")
+    for i in range(out_res.shape[0]):
+      for j in range(out_lens_res[i]):
+        assert np.alltrue(out_res[i, j] == labels_res[i, j]), \
+          "Position (%s,%s) not gathered correctly!" % (i, j)
+        assert np.alltrue(out_res[i, j] == src_res[i, positions_res[i, j]]), \
+          "Position (%s,%s) not gathered correctly!" % (i, j)
+    print(u'\N{check mark}')
+
+
+def test_MLM():
+  # test masking on its own:
+  _test_ProbabilisticMaskLayer(with_prob=False, mask_factor=(0.0, 1.0))
+  _test_ProbabilisticMaskLayer(with_prob=True)
+  _test_ProbabilisticMaskLayer(with_prob=False)
+  _test_ProbabilisticMaskLayer(with_prob=False, n_feature=100)
+  _test_ProbabilisticMaskLayer(fixed_mask=True, mask_factor=1.0, with_prob=False)
+  # test gathering on its own:
+  _test_GatherPositionsLayer()
+  # test them together:
+  _test_mask_and_gather(with_prob=False)
+  _test_mask_and_gather(with_prob=True)
```

Comment:
You should make 3 separate test cases for these. You should not combine all into a single test case `test_MLM`.
(But anyway, see discussion above. Some of the layers probably should be removed.)